blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,323 @@
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+### Preconditions
+
+Sometimes the refinements used in a dependent function type can also be written conveniently as a "precondition", e.g.:
+
+```racket
+(: safe-vector-ref
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Integer])
+                #:pre (v n) (<= 0 n (- (vector-length v) 1))
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+This function type is equivalent to the previous version, but the refining proposition that requires that `n` be a valid index for `v` now directly references `n` and has been written directly after the list of arguments:
+
+```racket
+#:pre (v n) (<= 0 n (- (vector-length v) 1))
+```
+
+We can read the precondition specification above as follows: the precondition for the function depends on arguments `v` and `n` and it requires that Typed Racket can prove `(<= 0 n (- (vector-length v) 1))` would produce `#true` at all call sites.
+
+Typed Racket may be able to provide more detailed feedback when the arguments are of the correct type but the precondition cannot be proved:
+
+```
+Type Checker: could not apply function;
+ unable to prove precondition
+  precondition: (<= 2 (- (vector-length v) 1)) 
+  in: (safe-vector-ref (vector 0 1) 2)
+```
+
+Here, Typed Racket first performed type inference and confirmed that the arguments  `(vector 0 1)` and `2` had acceptable types before discovering the precondition was unprovable at the call site.
+
+We believe the error messages produced when using preconditions may be more informative in general. Your mileage may vary.
+
+## Programming with refinement types
+
+_Note: these examples should work on any Racket builds from 3 November or later.
+To use with Racket v6.11, an additional small snippet of code (included at the end of
+this post) is required._
+
+When programming with refinement types, use the `#:with-refinements` option on the `#lang` line to tell Typed Racket to track additional logical information about program terms:
+
+```racket
+;; safe-vector-ops.rkt
+#lang typed/racket/base #:with-refinements
+```
+
+In particular, this will tell Typed Racket to remember that the vector produced by `(make-vector n ...)` has length `n`, which will be essential for typing our vector operations.
+
+Now we can also declare more detailed types for `vector-set!` and `vector-ref` since these more specific types are subtypes of their more permissive standard types. If we're feeling adventurous, we may use the `unsafe-` variants from `racket/unsafe/ops` that do not check their bounds, since Typed Racket will be proving they are never used with invalid indices:
+
+```racket
+(require (only-in racket/unsafe/ops unsafe-vector-ref
+                                    unsafe-vector-set!))
```

Comment:
Wait, this sounds backwards. "You can write this type if TR can prove something about all call site".

Suggestion:
> the precondition for `safe-vector-ref` depends on arguments `v` and `n` and states that an application of `safe-vector-ref` to a vector `v` and integer `n` is well typed only if `(<= 0 n (- (vector-length v) 1))`